polyply/src/gen_seq.py

```diff
@@ -0,0 +1,254 @@
+# Copyright 2020 University of Groningen
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import random
+import json
+import networkx as nx
+from networkx.readwrite import json_graph
+from .load_library import load_library
+from .generate_templates import find_atoms
+
+
+def _branched_graph(resname, branching_f, n_levels):
+    """
+    Generate a tree graph of branching factor `branching_f`
+    and number of generations as `n_levels` - 1. Each node
+    gets the attribute resname.
+
+    Parameters
+    ----------
+    resname:  str
+    branching_f: int
+    n_levels: int
+
+    Returns
+    -------
+    `:class:networkx.Graph`
+    """
+    graph = nx.balanced_tree(r=branching_f, h=n_levels-1)
+    resnames = {idx: resname for idx in graph.nodes}
+    nx.set_node_attributes(graph, resnames, "resname")
+    return graph
+
+
+def _random_replace_nodes(graph, residues, weights):
+    """
+    Randomly replace resname attribute of `graph`
+    based on the names in `residues` taking into
+    account `weights` provided.
+
+    Parameters
+    ----------
+    graph: `:class:networkx.Graph`
+    residues: list[str]
+    weights: list[float]
+
+    Returns:
+    --------
+    `:class:networkx.Graph`
+    """
+    for node in graph.nodes:
+        resname = random.choices(residues, weights=weights)
+        graph.nodes[node]["resname"] = resname[0]
```

Comment:
Consider passing "resname" as (default) argument

Reply:
how do you mean?

Reply:
give the function an argument `attribute` (or something), which by default is `'resname'` Things like that help in making functions reusable.